### ISSUE TITLE
Binlogs are incomplete with `MSBuildDebugEngine=1`

### ISSUE THREAD

**Issue by mawosoft** [mawosoft/Mawosoft.Extensions.BenchmarkDotNet#146]
Binlogs created via `MSBuildDebugEngine = 1` are incomplete. Will probably be fixed in MSBuild 17.4.
See dotnet/msbuild/issues/7780, dotnet/msbuild/pull/7897
Either revert to using -bl switch or check if adding additional env vars is sufficient.

**Comment by mawosoft** [comment]
## Mimic MSBuild -bl switch via environment vars

### TL;DR

`MSBuildDebugEngine` alone produces incomplete binlogs in MSBuild <=17.3
See dotnet/msbuild/issues/7780, dotnet/msbuild/pull/7897
Use the following env vars to fill in the blanks:
```
MSBuildDebugEngine=1
MSBUILDLOGTASKINPUTS=1
MSBUILDTARGETOUTPUTLOGGING=true
MSBUILDLOGIMPORTS=1                    # has only very subtle effect
MSBUILDLOGALLENVIRONMENTVARIABLES=true # Required in MSBuild >=17.4 to log *all* env vars
```

### Research setup

- Folder: C:\Users\mw\Projects\Mawosoft.Extensions.BenchmarkDotNet\src\Mawosoft.Extensions.BenchmarkDotNet\
- Cmd: dotnet build --no-restore
  This has been restored and build before, thus all of these are "build again", so to speak.

Log file          | settings
------------------|------------------
bl.binlog         | -bl switch
engine.binlog     | $env:MSBuildDebugEngine=1
output.binlog     | $env:MSBuildDebugEngine=1, $env:MSBUILDTARGETOUTPUTLOGGING='true'
out_imp.binlog    | $env:MSBuildDebugEngine=1, $env:MSBUILDTARGETOUTPUTLOGGING='true', $env:MSBUILDLOGIMPORTS=1
out_in.binlog     | $env:MSBuildDebugEngine=1, $env:MSBUILDTARGETOUTPUTLOGGING='true', $env:MSBUILDLOGTASKINPUTS=1
out_in_imp.binlog | $env:MSBuildDebugEngine=1, $env:MSBUILDTARGETOUTPUTLOGGING='true', $env:MSBUILDLOGTASKINPUTS=1, $env:MSBUILDLOGIMPORTS=1

### Diffs

#### bl vs. engine (MSBuildDebugEngine)
- Environment
```diff
+MSBuildDebugEngine = 1
+MSBUILDDEBUGPATH = ..path.. # auto-created if not set with MSBuildDebugEngine
-MSBUILDLOGIMPORTS = 1              # auto-created by -bl
-MSBUILDTARGETOUTPUTLOGGING = true  # auto-created by -bl
```
- Evaluation
No diffs except env (also listed here). SDK imports available in both.
- Build
Subtree lines: 32927 vs 1881
```diff
- Task Parameters (The Parameters node is missing, but some tasks seem to log an (in) param 'value', (out) param 'value' list or even custom nodes.)
- Task OutputProperties
- Task OutputItems
- Target property definitions
- Target item definitions
- TargetOutputs (Target Returns)
```

#### bl vs. output (MSBuildDebugEngine, MSBUILDTARGETOUTPUTLOGGING)
- Environment
```diff
+MSBuildDebugEngine = 1
+MSBUILDDEBUGPATH = ..path.. # auto-created if not set with MSBuildDebugEngine
-MSBUILDLOGIMPORTS = 1              # auto-created by -bl
```
- Evaluation
No diffs except env (also listed here). SDK imports available in both.
- Build
Subtree lines: 32927 vs 6409
```diff
- Task Parameters (The Parameters node is missing, but some tasks seem to log an (in) param 'value', (out) param 'value' list or even custom nodes.)
- Task OutputProperties
- Task OutputItems
- Target property definitions
- Target item definitions
= TargetOutputs (Target Returns)
```

#### bl vs. out_imp (MSBuildDebugEngine, MSBUILDTARGETOUTPUTLOGGING, MSBUILDLOGIMPORTS)
- Environment
```diff
+MSBuildDebugEngine = 1
+MSBUILDDEBUGPATH = ..path.. # auto-created if not set with MSBuildDebugEngine
```
- Evaluation, Build
Same as output.binlog. `MSBUILDLOGIMPORTS` seems not to have any effect.

#### bl vs. out_in (MSBuildDebugEngine, MSBUILDTARGETOUTPUTLOGGING, MSBUILDLOGTASKINPUTS)
- Environment
```diff
+MSBuildDebugEngine = 1
+MSBUILDDEBUGPATH = ..path.. # auto-created if not set with MSBuildDebugEngine
-MSBUILDLOGIMPORTS = 1              # auto-created by -bl
+MSBUILDLOGTASKINPUTS=1
```
- Evaluation
No diffs except env (also listed here). SDK imports available in both.
- Build
Subtree lines: 32927 vs 33335
```diff
+ Target ResolveAssemblyReferences -> Task ResolveAssemblyReference
+        Inputs node duplicates Parameters and also includes unused parameters (not set, null, '')
- Target _GetCopyToOutputDirectoryItemsFromThisProject
-        TargetOutputs have an addtl. item metadata MSBuildSourceTargetName = _GetCopyToOutputDirectoryItemsFromThisProject
= Task Parameters
= Task OutputProperties
= Task OutputItems
= Target property definitions
= Target item definitions
= TargetOutputs (Target Returns)
```

#### bl vs. out_in_imp (MSBuildDebugEngine, MSBUILDTARGETOUTPUTLOGGING, MSBUILDLOGTASKINPUTS, MSBUILDLOGIMPORTS)
- Environment
```diff
+MSBuildDebugEngine = 1
+MSBUILDDEBUGPATH = ..path.. # auto-created if not set with MSBuildDebugEngine
+MSBUILDLOGTASKINPUTS=1
```
- Evaluation
No diffs except env (also listed here). SDK imports available in both.
- Build
Subtree lines: 32927 vs 33345
```diff
+ Target ResolveAssemblyReferences -> Task ResolveAssemblyReference
+        Inputs node duplicates Parameters and also includes unused parameters (not set, null, '')
- Target _GetCopyToOutputDirectoryItemsFromThisProject
-        TargetOutputs have an addtl. item metadata MSBuildSourceTargetName = _GetCopyToOutputDirectoryItemsFromThisProject
-        In this variant only some of those metadata are missing
= Task Parameters
= Task OutputProperties
= Task OutputItems
= Target property definitions
= Target item definitions
= TargetOutputs (Target Returns)
```
